In [1]:
from consolidation.plateConsol import EchotoPlateConsol
from consolidation.plateConsol import CGEplatesConsol
from consolidation.ToDatabase import GB
from consolidation.ToDatabase.GB import ProjectScraper

from neo4j import GraphDatabase

import pandas as pd
import numpy as np 
from datetime import datetime
import os
import re

In [2]:
DIRS = r'test/1_Agar CP/'
FirstMeta = EchotoPlateConsol.PlateConverter(DIRS).convert()

In [3]:
FirstMeta.head(n=5)

ECHO_Source Plate Name ECHO_Source Plate Barcode ECHO_Source Well  \
0        Source Plate[1]                 Source[1]               B1   
1        Source Plate[1]                 Source[1]               B1   
2        Source Plate[1]                 Source[1]               B1   
3        Source Plate[1]                 Source[1]               B1   
4        Source Plate[1]                 Source[1]               B1   

  384_AGAR_PLATE echo_well_pick_actionTime 384_AGAR_PLATE_WELL  \
0       A0000404       09/05/2023 16:19:40                  O6   
1       A0000404       09/05/2023 16:19:40                  M6   
2       A0000404       09/05/2023 16:19:40                  J6   
3       A0000404       09/05/2023 16:19:40                  O5   
4       A0000404       09/05/2023 16:19:40                  L4   

  96_GROWTH_PLATE Sample Name                                log_file  \
0        G0000426          B7  UnifiedTransfer09-05-2023 16.18.38.log   
1        G0000426          C7  UnifiedTransfer09-05-2023 16.18.38.log   
2        G0000426          D7  UnifiedTransfer09-05-2023 16.18.38.log   
3        G0000426          E7  UnifiedTransfer09-05-2023 16.18.38.log   
4        G0000426          F7  UnifiedTransfer09-05-2023 16.18.38.log   

  EchoCherryPick_well_pick_actionTime EchoCherryPick_method_start_time  \
0                 2023-09-05 16:19:40              2023-09-05 16:18:38   
1                 2023-09-05 16:19:40              2023-09-05 16:18:38   
2                 2023-09-05 16:19:40              2023-09-05 16:18:38   
3                 2023-09-05 16:19:40              2023-09-05 16:18:38   
4                 2023-09-05 16:19:40              2023-09-05 16:18:38   

  Autiomation_SerialNum  
0           B8758519K13  
1           B8758519K13  
2           B8758519K13  
3           B8758519K13  
4           B8758519K13

In [4]:
dp = r'test/3_Harvest/'

ToDPplate = EchotoPlateConsol.DilutionPlateCreator(dp, FirstMeta).create_dilution_plates()

ToDPplate.head(n=2)

ECHO_Source Plate Name ECHO_Source Plate Barcode ECHO_Source Well  \
0        Source Plate[1]                 Source[1]               B1   
1        Source Plate[1]                 Source[1]               B1   

  384_AGAR_PLATE echo_well_pick_actionTime 384_AGAR_PLATE_WELL  \
0       A0000404       09/05/2023 16:19:40                  O6   
1       A0000404       09/05/2023 16:19:40                  M6   

  96_GROWTH_PLATE Sample Name                                log_file  \
0        G0000426          B7  UnifiedTransfer09-05-2023 16.18.38.log   
1        G0000426          C7  UnifiedTransfer09-05-2023 16.18.38.log   

  EchoCherryPick_well_pick_actionTime EchoCherryPick_method_start_time  \
0                 2023-09-05 16:19:40              2023-09-05 16:18:38   
1                 2023-09-05 16:19:40              2023-09-05 16:18:38   

  Autiomation_SerialNum Dilution_Plate                      Dilution_Log  \
0           B8758519K13       D0000002  Pipetting09-07-2023 13.15.04.log   
1           B8758519K13       D0000002  Pipetting09-07-2023 13.15.04.log   

  Dilution_Autiomation_Serial_Number  
0                        B8758519K13  
1                        B8758519K13

In [5]:
gp = r'test/4_GLY'

In [6]:
ToGP = EchotoPlateConsol.GlycerolPlates(gp, ToDPplate).convert()

In [7]:
ToGP.head(n=3)

ECHO_Source Plate Name ECHO_Source Plate Barcode ECHO_Source Well  \
0        Source Plate[1]                 Source[1]               B1   
1        Source Plate[1]                 Source[1]               B1   
2        Source Plate[1]                 Source[1]               B1   

  384_AGAR_PLATE echo_well_pick_actionTime 384_AGAR_PLATE_WELL  \
0       A0000404       09/05/2023 16:19:40                  O6   
1       A0000404       09/05/2023 16:19:40                  M6   
2       A0000404       09/05/2023 16:19:40                  J6   

  96_GROWTH_PLATE Sample Name                                log_file  \
0        G0000426          B7  UnifiedTransfer09-05-2023 16.18.38.log   
1        G0000426          C7  UnifiedTransfer09-05-2023 16.18.38.log   
2        G0000426          D7  UnifiedTransfer09-05-2023 16.18.38.log   

  EchoCherryPick_well_pick_actionTime EchoCherryPick_method_start_time  \
0                 2023-09-05 16:19:40              2023-09-05 16:18:38   
1                 2023-09-05 16:19:40              2023-09-05 16:18:38   
2                 2023-09-05 16:19:40              2023-09-05 16:18:38   

  Autiomation_SerialNum Dilution_Plate                      Dilution_Log  \
0           B8758519K13       D0000002  Pipetting09-07-2023 13.15.04.log   
1           B8758519K13       D0000002  Pipetting09-07-2023 13.15.04.log   
2           B8758519K13       D0000002  Pipetting09-07-2023 13.15.04.log   

  Dilution_Autiomation_Serial_Number Glycerol_Plate  \
0                        B8758519K13     GLY0000002   
1                        B8758519K13     GLY0000002   
2                        B8758519K13     GLY0000002   

  Glycerol_Autiomation_Serial_Number  
0                        B8758519K13  
1                        B8758519K13  
2                        B8758519K13

In [8]:
#--directory to the dilution logs 
dp = r'test/5_Decant'

In [9]:
#--initiate the getDilutionPlate class
FORCGE = CGEplatesConsol.GetDilutionPlates(ToGP, dp)

In [10]:
#--check dilution plates
FORCGE.getDplates()
FORCGE.Dplates

array(['D0000002', 'D0000004', 'D0000006', 'D0000008', 'D0000010',
       'D0000012', 'D0000014', 'D0000016'], dtype=object)

In [11]:
FORCGE.getCGEPlates()

In [12]:
FORCGE.sol

{'D0000002': 'S0000002', 'D0000010': 'S0000010'}

In [13]:
FORCGE.insol

{'D0000002': 'I0000002', 'D0000010': 'I0000010'}

In [14]:
#--Convert Insol plates
ToGP['Insol_Plate'] = ToGP['Dilution_Plate'].apply(lambda x: FORCGE.Analyzed_Plates_ISol(x)) 

In [15]:
ToGP.head(n=3)

ECHO_Source Plate Name ECHO_Source Plate Barcode ECHO_Source Well  \
0        Source Plate[1]                 Source[1]               B1   
1        Source Plate[1]                 Source[1]               B1   
2        Source Plate[1]                 Source[1]               B1   

  384_AGAR_PLATE echo_well_pick_actionTime 384_AGAR_PLATE_WELL  \
0       A0000404       09/05/2023 16:19:40                  O6   
1       A0000404       09/05/2023 16:19:40                  M6   
2       A0000404       09/05/2023 16:19:40                  J6   

  96_GROWTH_PLATE Sample Name                                log_file  \
0        G0000426          B7  UnifiedTransfer09-05-2023 16.18.38.log   
1        G0000426          C7  UnifiedTransfer09-05-2023 16.18.38.log   
2        G0000426          D7  UnifiedTransfer09-05-2023 16.18.38.log   

  EchoCherryPick_well_pick_actionTime EchoCherryPick_method_start_time  \
0                 2023-09-05 16:19:40              2023-09-05 16:18:38   
1                 2023-09-05 16:19:40              2023-09-05 16:18:38   
2                 2023-09-05 16:19:40              2023-09-05 16:18:38   

  Autiomation_SerialNum Dilution_Plate                      Dilution_Log  \
0           B8758519K13       D0000002  Pipetting09-07-2023 13.15.04.log   
1           B8758519K13       D0000002  Pipetting09-07-2023 13.15.04.log   
2           B8758519K13       D0000002  Pipetting09-07-2023 13.15.04.log   

  Dilution_Autiomation_Serial_Number Glycerol_Plate  \
0                        B8758519K13     GLY0000002   
1                        B8758519K13     GLY0000002   
2                        B8758519K13     GLY0000002   

  Glycerol_Autiomation_Serial_Number Insol_Plate  
0                        B8758519K13    I0000002  
1                        B8758519K13    I0000002  
2                        B8758519K13    I0000002

In [16]:
ToGP['Sol_Plate'] = ToGP['Dilution_Plate'].apply(lambda x: FORCGE.Analyzed_Plates_Sol(x))

In [17]:
ToGP.head(n=3)

ECHO_Source Plate Name ECHO_Source Plate Barcode ECHO_Source Well  \
0        Source Plate[1]                 Source[1]               B1   
1        Source Plate[1]                 Source[1]               B1   
2        Source Plate[1]                 Source[1]               B1   

  384_AGAR_PLATE echo_well_pick_actionTime 384_AGAR_PLATE_WELL  \
0       A0000404       09/05/2023 16:19:40                  O6   
1       A0000404       09/05/2023 16:19:40                  M6   
2       A0000404       09/05/2023 16:19:40                  J6   

  96_GROWTH_PLATE Sample Name                                log_file  \
0        G0000426          B7  UnifiedTransfer09-05-2023 16.18.38.log   
1        G0000426          C7  UnifiedTransfer09-05-2023 16.18.38.log   
2        G0000426          D7  UnifiedTransfer09-05-2023 16.18.38.log   

  EchoCherryPick_well_pick_actionTime EchoCherryPick_method_start_time  \
0                 2023-09-05 16:19:40              2023-09-05 16:18:38   
1                 2023-09-05 16:19:40              2023-09-05 16:18:38   
2                 2023-09-05 16:19:40              2023-09-05 16:18:38   

  Autiomation_SerialNum Dilution_Plate                      Dilution_Log  \
0           B8758519K13       D0000002  Pipetting09-07-2023 13.15.04.log   
1           B8758519K13       D0000002  Pipetting09-07-2023 13.15.04.log   
2           B8758519K13       D0000002  Pipetting09-07-2023 13.15.04.log   

  Dilution_Autiomation_Serial_Number Glycerol_Plate  \
0                        B8758519K13     GLY0000002   
1                        B8758519K13     GLY0000002   
2                        B8758519K13     GLY0000002   

  Glycerol_Autiomation_Serial_Number Insol_Plate Sol_Plate  
0                        B8758519K13    I0000002  S0000002  
1                        B8758519K13    I0000002  S0000002  
2                        B8758519K13    I0000002  S0000002

In [18]:
#--directory to the reduced CGE plates
reduced = r'test/6_CGE'

In [19]:
#--initiate the ReducedNonReducedPlateConverter class
FORCGEREDUCED = CGEplatesConsol.ReducedNonReducedPlateConverter(reduced)

In [20]:
#--collect and check the plates
FORCGEREDUCED.collectPlates()
FORCGEREDUCED.PLATES

['S0000002',
 'RED0000008',
 'S0000010',
 'RED0000016',
 'I0000002',
 'RED0000024',
 'I0000010',
 'RED0000032']

In [21]:
#--collect the reduced plates converter
FORCGEREDUCED.collectReducedPlates()

In [22]:
#--inspect the reduced plates
FORCGEREDUCED.SOL_REDUCED_PLATES

{'S0000002': 'RED0000008', 'S0000010': 'RED0000016'}

In [23]:
#--inspect the reduced plates insol
FORCGEREDUCED.INSOL_REDUCED_PLATES

{'I0000002': 'RED0000024', 'I0000010': 'RED0000032'}

In [24]:
#--create SOl and reduced CGE column
ToGP['Sol_Reduced_CGEPlate'] = ToGP['Sol_Plate'].apply(lambda x: FORCGEREDUCED.Analyzed_CGEPlates_Sol(x))
ToGP['Insol_Reduced_CGEPlate'] = ToGP['Insol_Plate'].apply(lambda x: FORCGEREDUCED.Analyzed_CGEPlates_ISol(x))
ToGP.head(n=3)

ECHO_Source Plate Name ECHO_Source Plate Barcode ECHO_Source Well  \
0        Source Plate[1]                 Source[1]               B1   
1        Source Plate[1]                 Source[1]               B1   
2        Source Plate[1]                 Source[1]               B1   

  384_AGAR_PLATE echo_well_pick_actionTime 384_AGAR_PLATE_WELL  \
0       A0000404       09/05/2023 16:19:40                  O6   
1       A0000404       09/05/2023 16:19:40                  M6   
2       A0000404       09/05/2023 16:19:40                  J6   

  96_GROWTH_PLATE Sample Name                                log_file  \
0        G0000426          B7  UnifiedTransfer09-05-2023 16.18.38.log   
1        G0000426          C7  UnifiedTransfer09-05-2023 16.18.38.log   
2        G0000426          D7  UnifiedTransfer09-05-2023 16.18.38.log   

  EchoCherryPick_well_pick_actionTime  ... Autiomation_SerialNum  \
0                 2023-09-05 16:19:40  ...           B8758519K13   
1                 2023-09-05 16:19:40  ...           B8758519K13   
2                 2023-09-05 16:19:40  ...           B8758519K13   

  Dilution_Plate                      Dilution_Log  \
0       D0000002  Pipetting09-07-2023 13.15.04.log   
1       D0000002  Pipetting09-07-2023 13.15.04.log   
2       D0000002  Pipetting09-07-2023 13.15.04.log   

  Dilution_Autiomation_Serial_Number Glycerol_Plate  \
0                        B8758519K13     GLY0000002   
1                        B8758519K13     GLY0000002   
2                        B8758519K13     GLY0000002   

  Glycerol_Autiomation_Serial_Number Insol_Plate Sol_Plate  \
0                        B8758519K13    I0000002  S0000002   
1                        B8758519K13    I0000002  S0000002   
2                        B8758519K13    I0000002  S0000002   

  Sol_Reduced_CGEPlate Insol_Reduced_CGEPlate  
0           RED0000008             RED0000024  
1           RED0000008             RED0000024  
2           RED0000008             RED0000024  

[3 rows x 21 columns]

In [25]:
CGEDATAA = CGEplatesConsol.GetCGEdata(r'test/OneDrive_2023-09-19/Peak Tables/')

In [26]:
#--Return only consolidated data frame
CGEDATAA.PLATE_DATA

Well Number Sample Name Well Label  Migration Time(sec) - Center  \
0               2          A1        A01                     29.902573   
1               7          A6        A06                     29.449489   
2               8          A7        A07                     29.417011   
3              10          A9        A09                     29.902010   
4              11         A10        A10                     29.848359   
...           ...         ...        ...                           ...   
1311           84          G7        G07                     25.500300   
1312           85          G8        G08                     25.443679   
1313           93          H4        H04                     25.347152   
1314           97          H8        H08                     25.334846   
1315           99         H10        H10                     25.312142   

      Peak Number      Height   Corr. Area  Expected Protein [kDa]  \
0              41   18.171939    22.117497                     150   
1              44   12.437826    12.021422                     150   
2              46   11.669323    11.170058                     150   
3              45   12.885283    11.762757                     150   
4              41  295.054164   118.888745                     150   
...           ...         ...          ...                     ...   
1311           26   19.659739    24.480091                      92   
1312           31  980.488958  1058.426180                      92   
1313           31  612.137908   643.944198                      92   
1314           30  492.578833   510.119262                      92   
1315           30  192.913822   231.023585                      92   

      Size [kDa]  Conc. (ng/ul)   % Purity   Type  FWHM (sec)  Comment  \
0     153.552518       8.219369   0.788827  EP150    0.269695      NaN   
1     146.610274       5.042328   1.141297  EP150    0.209525      NaN   
2     146.115883       4.836147   1.035069  EP150    0.202330      NaN   
3     153.543838       5.314913   0.596501  EP150    0.193818      NaN   
4     152.717501      53.760758   5.867834  EP150    0.073671      NaN   
...          ...            ...        ...    ...         ...      ...   
1311   91.857635      19.828129   5.607666   EP92    0.336963      NaN   
1312   91.091156     854.423686  54.443208   EP92    0.218054      NaN   
1313   91.430113     529.657993  44.302174   EP92    0.206831      NaN   
1314   91.261380     420.331287  35.266231   EP92    0.198863      NaN   
1315   90.950401     189.572942  23.464912   EP92    0.202843      NaN   

                          Plate Name Reduced_Plate  
0     RED0000012_2023-09-11_11-53-19    RED0000012  
1     RED0000012_2023-09-11_11-53-19    RED0000012  
2     RED0000012_2023-09-11_11-53-19    RED0000012  
3     RED0000012_2023-09-11_11-53-19    RED0000012  
4     RED0000012_2023-09-11_11-53-19    RED0000012  
...                              ...           ...  
1311  RED0000038_2023-09-12_13-24-35    RED0000038  
1312  RED0000038_2023-09-12_13-24-35    RED0000038  
1313  RED0000038_2023-09-12_13-24-35    RED0000038  
1314  RED0000038_2023-09-12_13-24-35    RED0000038  
1315  RED0000038_2023-09-12_13-24-35    RED0000038  

[1316 rows x 16 columns]

In [27]:
#--save consolidated CGE data Frame
CGEDATAA.PLATE_DATA.to_csv('test/OneDrive_2023-09-19/Peak Tables/Consolidated_CGE.csv', index=False)

In [28]:
#--check dictionary 
CGEDATAA.reducedPDATA

{'RED0000012': {'A1': (8.21936898690468,
   'RED0000012_2023-09-11_11-53-19_PeakTable.csv'),
  'A6': (5.04232783245498, 'RED0000012_2023-09-11_11-53-19_PeakTable.csv'),
  'A7': (4.83614688329577, 'RED0000012_2023-09-11_11-53-19_PeakTable.csv'),
  'A9': (5.31491286910658, 'RED0000012_2023-09-11_11-53-19_PeakTable.csv'),
  'A10': (53.7607579135201, 'RED0000012_2023-09-11_11-53-19_PeakTable.csv'),
  'A11': (5.31966626800305, 'RED0000012_2023-09-11_11-53-19_PeakTable.csv'),
  'B1': (8.00901855016116, 'RED0000012_2023-09-11_11-53-19_PeakTable.csv'),
  'B2': (4.6049771141276, 'RED0000012_2023-09-11_11-53-19_PeakTable.csv'),
  'B9': (6.15874372215905, 'RED0000012_2023-09-11_11-53-19_PeakTable.csv'),
  'B10': (7.94418020911716, 'RED0000012_2023-09-11_11-53-19_PeakTable.csv'),
  'C1': (13.2281036620326, 'RED0000012_2023-09-11_11-53-19_PeakTable.csv'),
  'C2': (5.75209559579097, 'RED0000012_2023-09-11_11-53-19_PeakTable.csv'),
  'C5': (6.59550866935967, 'RED0000012_2023-09-11_11-53-19_PeakTable.

In [29]:
#--get entire META dataframe
ToGP["Sol Conc. (ng/ul)"] = ToGP.apply(lambda row: CGEDATAA.get_peak_data(row['Sol_Reduced_CGEPlate'], row['Sample Name']), axis=1)
ToGP["Insol Conc. (ng/ul)"] = ToGP.apply(lambda row: CGEDATAA.get_peak_data(row['Insol_Reduced_CGEPlate'], row['Sample Name']), axis=1)
ToGP["Peak Log"] = ToGP.apply(lambda row: CGEDATAA.get_peak_log(row['Sol_Reduced_CGEPlate'], row['Sample Name']), axis=1)

In [30]:
ToGP.head(n=3)

ECHO_Source Plate Name ECHO_Source Plate Barcode ECHO_Source Well  \
0        Source Plate[1]                 Source[1]               B1   
1        Source Plate[1]                 Source[1]               B1   
2        Source Plate[1]                 Source[1]               B1   

  384_AGAR_PLATE echo_well_pick_actionTime 384_AGAR_PLATE_WELL  \
0       A0000404       09/05/2023 16:19:40                  O6   
1       A0000404       09/05/2023 16:19:40                  M6   
2       A0000404       09/05/2023 16:19:40                  J6   

  96_GROWTH_PLATE Sample Name                                log_file  \
0        G0000426          B7  UnifiedTransfer09-05-2023 16.18.38.log   
1        G0000426          C7  UnifiedTransfer09-05-2023 16.18.38.log   
2        G0000426          D7  UnifiedTransfer09-05-2023 16.18.38.log   

  EchoCherryPick_well_pick_actionTime  ... Dilution_Autiomation_Serial_Number  \
0                 2023-09-05 16:19:40  ...                        B8758519K13   
1                 2023-09-05 16:19:40  ...                        B8758519K13   
2                 2023-09-05 16:19:40  ...                        B8758519K13   

  Glycerol_Plate Glycerol_Autiomation_Serial_Number Insol_Plate Sol_Plate  \
0     GLY0000002                        B8758519K13    I0000002  S0000002   
1     GLY0000002                        B8758519K13    I0000002  S0000002   
2     GLY0000002                        B8758519K13    I0000002  S0000002   

  Sol_Reduced_CGEPlate Insol_Reduced_CGEPlate Sol Conc. (ng/ul)  \
0           RED0000008             RED0000024              None   
1           RED0000008             RED0000024          5.655277   
2           RED0000008             RED0000024              None   

  Insol Conc. (ng/ul)                                      Peak Log  
0                None                                          None  
1                None  RED0000008_2023-09-07_15-10-58_PeakTable.csv  
2                None                                          None  

[3 rows x 24 columns]

In [31]:
ToGP.to_csv('test/Consolidated.csv', index=None)

## Define the database structure

* EchoPlates (ID's and barcodes for echo plates)
* 384 Agar Plates (ID's and barcodes for 384 agar plates)
* 96 well growth plates (ID's and barcodes for 96 well growth plates)
* Dilution plates (ID's and barcodes for 96 well dilution plates)
* Glycerol plates (ID's and barcodes for 96 well dilution plates)
* Soluble fraction plates (ID's and barcodes for 96 well soluble fraction plates)
* Insoluble fraction plates (ID's and barcodes for 96 well plates)
* Reduced plates (ID's and barcodes for CGE plates)

In [1]:
import consolidation.ToDatabase.PlateGen as pcbaby 
# Define the connection details
uri = "bolt://localhost:7687"  # Update this with your Neo4j server URI
username = "neo4j"
password = 'plates1234'

In [2]:
#--get EchoPlates
ECHONODES = pcbaby.getPlateDB("EchoPlate", "Echo Plate")

In [3]:
#--check nodes
ECHONODES.node_id, ECHONODES.plate_type

('EchoPlate', 'Echo Plate')

In [4]:
#--fetech echo plates
ECHONODES.get_plates_property(ECHONODES.node_id, ECHONODES.plate_type, uri, username, password)

In [5]:
#--return Echo Plates in catalogue
ECHONODES.PlateCatalogue

['R754166', 'R603478', 'R901485']

In [6]:
#--create new plates
ECHONODES_EXTENDED = pcbaby.generate_random_numbers_extensions(4, "R", ECHONODES.PlateCatalogue).generate_random_numbers()

In [7]:
#--check plates
ECHONODES_EXTENDED

['R754166', 'R603478', 'R901485', 'R300059', 'R837092', 'R853434', 'R656353']

In [4]:
# Define the connection details
uri = "bolt://localhost:7687"  # Update this with your Neo4j server URI
username = "neo4j"
password = 'fastassscreen'

In [5]:
#--test connection
GB.Neo4jConnector(uri, username, password).connect().test_connection()

'Connected'

In [6]:
#--basic project MetaData
projectInfo = pd.read_csv(r'../ProjectInfo.csv')

In [7]:
#--return basic meta information
Project, START, END, SOW, GENES, PROTEINNAME, GENESeq, GENE_LeaderPROTEIN, Gene_NoLeaderProtein = ProjectScraper(projectInfo).scrape_project_input()

In [10]:
#--push to graphical database
with GraphDatabase.driver(uri, auth=(username, password)) as driver:
    with driver.session() as session:
        PROTEINLEADERINFO = GENE_LeaderPROTEIN.get(GENES[0])
        PROTEINNLINFO = Gene_NoLeaderProtein.get(GENES[0])
        session.execute_write(GB.create_project, Project[0], START, END, SOW)
        session.execute_write(GB.create_gene_to_projects, Project[0], GENES[0], GENESeq,
                            PROTEINNAME,  PROTEINLEADERINFO[1], PROTEINLEADERINFO[2], PROTEINLEADERINFO[3], PROTEINLEADERINFO[4],
                            PROTEINNLINFO[1], PROTEINNLINFO[2], PROTEINNLINFO[3], PROTEINNLINFO[4], PROTEINLEADERINFO[0],  PROTEINNLINFO[0])